feat: HMAC SHA256 Authentication (#795)
codeigniter4/shield: docs/authentication.md

    // do something....
}
```

## HMAC SHA256 Token Authenticator

> The HMAC-SHA256 authenticator supports the use of revoke-able API keys without using OAuth. This provides

Reviewer: ```suggestion
The HMAC-SHA256 authenticator supports the use of revocable API keys without using OAuth. This provides
```

an alternative to a token that is passed in every request and instead uses a shared secret that is used to sign
the request in a secure manner. Like authorization tokens, these are commonly used to provide third-party developers
access to your API. These keys typically have a very long expiration time, often years.

These are also suitable for use with mobile applications. In this case, the user would register/sign-in
with their email/password. The application would create a new access token for them, with a recognizable
name, like John's iPhone 12, and return it to the mobile application, where it is stored and used
in all future requests.

> **NOTE:** for the purpose of this documentation and to maintain a level of consistency with the Authorization Tokens,
the term "Token" will be used to represent a set of API Keys (key and secretKey).

> ### Usage

Reviewer 2: Add an empty line.

In order to use HMAC Keys/Token the `Authorization` header will be set to the following in the request:

```

> Authorization: HMAC-SHA256 <key>:<HMAC HASH of request body>

Reviewer: ```suggestion
Authorization: HMAC-SHA256 <key>:<HMAC-HASH-of-request-body>
```

```

The code to do this will look something like this:

```php
<?php

> header("Authorization: HMAC-SHA256 {$key}:" . hash_hmac('sha256', $requestBody, $secretKey));

Reviewer: Can you add a usable example where `$key`, `$requestBody`, and `$secretKey` are defined. I'm fearing someone would copy-paste this example and complain it's not working.

Reviewer 2: CI4 users should not use `header()` at all. Or at least I think it is not good practice.
If you fear someone would copy-paste, the code should use the Response class?

Author: @paulbalandan I like that idea and will add a tangible example shortly.

@kenjis I used `header()` here as it is intended to be a generic reference for a third-party app calling this API.  I felt that a CI4 CURLRequest class reference here may be confusing as to the purpose of this code sample.  If you feel strongly otherwise, I will adjust it.

Author: I added a second example using the CURLRequest() class and showing an actual hash calculation.

```

### HMAC Keys/API Authentication

Using HMAC keys requires that you either use/extend `CodeIgniter\Shield\Models\UserModel` or
use the `CodeIgniter\Shield\Authentication\Traits\HasHmacTokens` on your own user model. This trait
provides all the custom methods needed to implement HMAC keys in your application. The necessary
database table, `auth_identities`, is created in Shield's only migration class, which must be run
before first using any of the features of Shield.

### Generating Access Keys

Access keys/tokens are created through the `generateHmacToken()` method on the user. This takes a name to
give to the token as the first argument. The name is used to display it to the user, so they can
differentiate between multiple tokens.

```php
$token = $user->generateHmacToken('Work Laptop');
```

> This creates the keys/tokens using a cryptographically secure random string. The keys opporate as shared keys.

Reviewer: ```suggestion
This creates the keys/tokens using a cryptographically secure random string. The keys operate as shared keys.
```

This means they are stored as-is in the database. The method returns an instance of
`CodeIgniters\Shield\Authentication\Entities\AccessToken`. The field `secret` is the 'key' the field `secret2` is

> the shared 'secredKey'. Both are required to when using this authentication method.

Reviewer: ```suggestion
the shared 'secretKey'. Both are required to when using this authentication method.
```

    // do something....
}
```

## HMAC SHA256 Token Authenticator

The HMAC-SHA256 authenticator supports the use of revocable API keys without using OAuth. This provides
an alternative to a token that is passed in every request and instead uses a shared secret that is used to sign
the request in a secure manner. Like authorization tokens, these are commonly used to provide third-party developers
access to your API. These keys typically have a very long expiration time, often years.

These are also suitable for use with mobile applications. In this case, the user would register/sign-in
with their email/password. The application would create a new access token for them, with a recognizable
name, like John's iPhone 12, and return it to the mobile application, where it is stored and used
in all future requests.

> **Note** For the purpose of this documentation, and to maintain a level of consistency with the Authorization Tokens,
> the term "Token" will be used to represent a set of API Keys (key and secretKey).

> ### Usage

Reviewer 2: ```suggestion
### Usage

```

In order to use HMAC Keys/Token the `Authorization` header will be set to the following in the request:

```
Authorization: HMAC-SHA256 <key>:<HMAC-HASH-of-request-body>
```

The code to do this will look something like this:

```php

> <?php
> 
> header("Authorization: HMAC-SHA256 {$key}:" . hash_hmac('sha256', $requestBody, $secretKey));

Reviewer 2: ```suggestion
header("Authorization: HMAC-SHA256 {$key}:" . hash_hmac('sha256', $requestBody, $secretKey));
```

```

> Using the CodeIgniter CURLRequest class:

Reviewer 2: ```suggestion

Using the CodeIgniter CURLRequest class:
```

```php
<?php

$client = \Config\Services::curlrequest();

$key = 'a6c460151b4cabbe1c1d73e08915ce8e';
$secretKey = '56c85232f0e5b55c05015476cd132c8d';

> $requestBody = '{"name":"John";"email":"john@example.com"}';

Author: Because I was working on this during a meeting... I will fix it. Thanks for the catch.

their use.** It is recommended that after that only the 'key' field is displayed to a user.  If a user loses the
'secretKey', they should be required to generate a new set of keys to use.

```php
$token = $user->generateHmacToken('Work Laptop');

echo 'Key: ' . $token->secret;
echo 'SecretKey: ' . $token->secret2;
```

### Revoking HMAC Keys

HMAC keys can be revoked through the `revokeHmacToken()` method. This takes the key as the only
argument. Revoking simply deletes the record from the database.

```php
$user->revokeHmacToken($key);
```

You can revoke all HMAC Keys with the `revokeAllHmacTokens()` method.

```php
$user->revokeAllHmacTokens();
```

### Retrieving HMAC Keys

The following methods are available to help you retrieve a user's HMAC keys:

```php
// Retrieve a set of HMAC Token/Keys by key
$token = $user->getHmacToken($key);

// Retrieve an HMAC token/keys by its database ID
$token = $user->getHmacTokenById($id);

// Retrieve all HMAC tokens as an array of AccessToken instances.
$tokens = $user->hmacTokens();
```

### HMAC Keys Lifetime

HMAC Keys/Tokens will expire after a specified amount of time has passed since they have been used.

> This uses the same configuration value as AccessTokens.

Author: That is a really good question.  I have it using the same configuration as I see usage being mutually exclusive, but if we want to allow someone to use both it perhaps should be separate.

> This uses the same configuration value as AccessTokens.

Reviewer 2: > I see usage being mutually exclusive

Good point. It seems in most cases.
Then leave it as it is, and let's consider if a request comes.

> By default, this is set to 1 year. You can change this value by setting the `accessTokenLifetime`

Reviewer 2: ```suggestion
By default, this is set to 1 year. You can change this value by setting the `$unusedTokenLifetime`
```

Author: The same issue exists in the Access Tokens section.  I have resolved it in both locations.
